Python 3.12 support

### Description

On calling pip install msgspec, I am getting the error 
`error C2039: 'ob_digit': is not a member of '_longobject'`
with all the versions of msgspec I've tried (0.13, 0.15.1, 0.16.0).

2023-07-03T15:25:18,464   msgspec\_core.c(152): error C2039: 'ob_digit': is not a member of '_longobject'
2023-07-03T15:25:18,464   msgspec\_core.c(159): error C2039: 'ob_digit': is not a member of '_longobject'

I've switched back to 3.11 in the meantime, this is just FYI.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -96,10 +96,16 @@ jobs:
       - name: Set up Environment
         if: github.event_name != 'release'
         run: |
-            echo "CIBW_SKIP=${CIBW_SKIP} *-musllinux_* cp38-*_aarch64 cp39-*_aarch64 cp311-*_aarch64" >> $GITHUB_ENV
+            echo "CIBW_SKIP=${CIBW_SKIP} *-musllinux_* cp38-*_aarch64 cp39-*_aarch64 cp311-*_aarch64 cp312-*_aarch64" >> $GITHUB_ENV
+
+      - name: Enable CPython 3.12 prerelease builds for Linux testing
+        if: github.event_name != 'release' && runner.os == 'Linux'
+        run: |
+          echo "CIBW_PRERELEASE_PYTHONS=True" >> $GITHUB_ENV
+          echo "CIBW_BUILD=cp38-* cp39-* cp310-* cp311-* cp312-*" >> $GITHUB_ENV
 
       - name: Build & Test Wheels
-        uses: pypa/cibuildwheel@v2.11.2
+        uses: pypa/cibuildwheel@v2.13.1
 
       - name: Upload artifact
         uses: actions/upload-artifact@v2

--- a/msgspec/_core.c
+++ b/msgspec/_core.c
@@ -153,10 +153,34 @@ ms_release_buffer(Py_buffer *view) {
  */
 static inline bool
 fast_long_extract_parts(PyObject *vv, bool *neg, uint64_t *scale) {
-    uint64_t prev, x = 0;
     PyLongObject *v = (PyLongObject *)vv;
+    uint64_t prev, x = 0;
+    bool negative;
+
+#if PY_VERSION_HEX >= 0x030c0000
+    /* CPython 3.12 changed int internal representation */
+    int sign = 1 - (v->long_value.lv_tag & _PyLong_SIGN_MASK);
+    negative = sign == -1;
+
+    if (_PyLong_IsCompact(v)) {
+        x = v->long_value.ob_digit[0];
+    }
+    else {
+        Py_ssize_t i = v->long_value.lv_tag >> _PyLong_NON_SIZE_BITS;
+        while (--i >= 0) {
+            prev = x;
+            x = (x << PyLong_SHIFT) + v->long_value.ob_digit[i];
+            if ((x >> PyLong_SHIFT) != prev) {
+                return true;
+            }
+        }
+        if (negative && x > (1ull << 63)) {
+            return true;
+        }
+    }
+#else
     Py_ssize_t i = Py_SIZE(v);
-    bool negative = false;
+    negative = false;
 
     if (MS_LIKELY(i == 1)) {
         x = v->ob_digit[0];
@@ -175,6 +199,7 @@ fast_long_extract_parts(PyObject *vv, bool *neg, uint64_t *scale) {
             return true;
         }
     }
+#endif
 
     *neg = negative;
     *scale = x;
@@ -1879,7 +1904,15 @@ PyTypeObject NoDefault_Type = {
     .tp_basicsize = 0
 };
 
+#if PY_VERSION_HEX >= 0x030c0000
+PyObject _NoDefault_Object = {
+    _PyObject_EXTRA_INIT
+    { _Py_IMMORTAL_REFCNT },
+    &NoDefault_Type
+};
+#else
 PyObject _NoDefault_Object = {1, &NoDefault_Type};
+#endif
 
 /*************************************************************************
  * UNSET singleton                                                       *
@@ -1975,7 +2008,16 @@ PyTypeObject Unset_Type = {
     .tp_basicsize = 0
 };
 
+#if PY_VERSION_HEX >= 0x030c0000
+PyObject _Unset_Object = {
+    _PyObject_EXTRA_INIT
+    { _Py_IMMORTAL_REFCNT },
+    &Unset_Type
+};
+#else
 PyObject _Unset_Object = {1, &Unset_Type};
+#endif
+
 
 /*************************************************************************
  * Factory                                                               *
@@ -11150,10 +11192,10 @@ mpack_encode_uncommon(EncoderState *self, PyTypeObject *type, PyObject *obj)
     else if (PyAnySet_Check(obj)) {
         return mpack_encode_set(self, obj);
     }
-    else if (PyDict_Contains(type->tp_dict, self->mod->str___dataclass_fields__)) {
+    else if (type->tp_dict != NULL && PyDict_Contains(type->tp_dict, self->mod->str___dataclass_fields__)) {
         return mpack_encode_object(self, obj);
     }
-    else if (PyDict_Contains(type->tp_dict, self->mod->str___attrs_attrs__)) {
+    else if (type->tp_dict != NULL && PyDict_Contains(type->tp_dict, self->mod->str___attrs_attrs__)) {
         return mpack_encode_object(self, obj);
     }
 
@@ -12048,10 +12090,10 @@ json_encode_uncommon(EncoderState *self, PyTypeObject *type, PyObject *obj) {
     else if (PyAnySet_Check(obj)) {
         return json_encode_set(self, obj);
     }
-    else if (PyDict_Contains(type->tp_dict, self->mod->str___dataclass_fields__)) {
+    else if (type->tp_dict != NULL && PyDict_Contains(type->tp_dict, self->mod->str___dataclass_fields__)) {
         return json_encode_object(self, obj);
     }
-    else if (PyDict_Contains(type->tp_dict, self->mod->str___attrs_attrs__)) {
+    else if (type->tp_dict != NULL && PyDict_Contains(type->tp_dict, self->mod->str___attrs_attrs__)) {
         return json_encode_object(self, obj);
     }
 
@@ -17739,10 +17781,10 @@ to_builtins(ToBuiltinsState *self, PyObject *obj, bool is_key) {
     else if (PyAnySet_Check(obj)) {
         return to_builtins_set(self, obj, is_key);
     }
-    else if (PyDict_Contains(type->tp_dict, self->mod->str___dataclass_fields__)) {
+    else if (type->tp_dict != NULL && PyDict_Contains(type->tp_dict, self->mod->str___dataclass_fields__)) {
         return to_builtins_object(self, obj);
     }
-    else if (PyDict_Contains(type->tp_dict, self->mod->str___attrs_attrs__)) {
+    else if (type->tp_dict != NULL && PyDict_Contains(type->tp_dict, self->mod->str___attrs_attrs__)) {
         return to_builtins_object(self, obj);
     }
     else if (self->enc_hook != NULL) {

--- a/tests/test_common.py
+++ b/tests/test_common.py
@@ -1913,7 +1913,7 @@ class TestTypedDict:
         if not hasattr(ns, "Required"):
             pytest.skip(f"{module}.Required is not available")
 
-        if not hasattr(ns.TypedDict("C"), "__required_keys__"):
+        if not hasattr(ns.TypedDict("C", {}), "__required_keys__"):
             # This should be Python 3.8, builtin typing only
             pytest.skip("partially optional TypedDict not supported")
 

--- a/tests/test_convert.py
+++ b/tests/test_convert.py
@@ -52,6 +52,7 @@ except ImportError:
 
 PY310 = sys.version_info[:2] >= (3, 10)
 PY311 = sys.version_info[:2] >= (3, 11)
+PY312 = sys.version_info[:2] >= (3, 12)
 
 uses_annotated = pytest.mark.skipif(Annotated is None, reason="Annotated not available")
 
@@ -883,6 +884,10 @@ class TestSequences:
             assert convert(1, out_type)
 
     @pytest.mark.parametrize("kind", ["list", "tuple", "fixtuple", "set"])
+    @pytest.mark.skipif(
+        PY312,
+        reason="CPython 3.12 internal changes prevent testing for recursion issues this way",
+    )
     def test_sequence_cyclic_recursion(self, kind):
         depth = 50
         if kind == "list":
@@ -1107,6 +1112,10 @@ class TestDict:
         with pytest.raises(ValidationError):
             convert(dictcls({"x": 1}), Dict[Tuple[int, int], int], str_keys=True)
 
+    @pytest.mark.skipif(
+        PY312,
+        reason="CPython 3.12 internal changes prevent testing for recursion issues this way",
+    )
     def test_dict_cyclic_recursion(self, dictcls):
         depth = 50
         typ = Dict[str, int]

